conn.go

```diff
@@ -454,251 +433,181 @@ func (c *Conn) deleteSession(s *Session) {
-				c.sessionsByChannelMu.RUnlock()
-				if !ok {
-					c.doneErr = fmt.Errorf("unexpected remote channel number %d", *body.RemoteChannel)
-					break
-				}
-
-				session.remoteChannel = fr.Channel
-				sessionsByRemoteChannel[fr.Channel] = session
-
-			case *frames.PerformEnd:
-				session, ok = sessionsByRemoteChannel[fr.Channel]
-				if !ok {
-					c.doneErr = fmt.Errorf("%T: didn't find channel %d in sessionsByRemoteChannel (PerformEnd)", fr.Body, fr.Channel)
-					break
-				}
-				// we MUST remove the remote channel from our map as soon as we receive
-				// the ack (i.e. before passing it on to the session mux) on the session
-				// ending since the numbers are recycled.
-				delete(sessionsByRemoteChannel, fr.Channel)
-
-			default:
-				// pass on performative to the correct session
-				session, ok = sessionsByRemoteChannel[fr.Channel]
-				if !ok {
-					c.doneErr = fmt.Errorf("%T: didn't find channel %d in sessionsByRemoteChannel", fr.Body, fr.Channel)
-				}
+		var (
+			session *Session
+			ok      bool
+		)
+
+		switch body := fr.Body.(type) {
+		// Server initiated close.
+		case *frames.PerformClose:
+			// connWriter will send the close performative ack on its way out.
+			// it's a SHOULD though, not a MUST.
+			debug.Log(3, "RX (connReader): %s", body)
+			if body.Error == nil {
+				return
 			}
+			err = body.Error
+			continue
 
+		// RemoteChannel should be used when frame is Begin
+		case *frames.PerformBegin:
+			if body.RemoteChannel == nil {
+				// since we only support remotely-initiated sessions, this is an error
+				// TODO: it would be ideal to not have this kill the connection
+				err = fmt.Errorf("%T: nil RemoteChannel", fr.Body)
+				continue
+			}
+			c.sessionsByChannelMu.RLock()
+			session, ok = c.sessionsByChannel[*body.RemoteChannel]
+			c.sessionsByChannelMu.RUnlock()
 			if !ok {
+				err = fmt.Errorf("unexpected remote channel number %d", *body.RemoteChannel)
 				continue
 			}
 
-			select {
-			case session.rx <- fr:
-			case <-c.closeMux:
-				return
+			session.remoteChannel = fr.Channel
+			sessionsByRemoteChannel[fr.Channel] = session
+
+		case *frames.PerformEnd:
+			session, ok = sessionsByRemoteChannel[fr.Channel]
+			if !ok {
+				err = fmt.Errorf("%T: didn't find channel %d in sessionsByRemoteChannel (PerformEnd)", fr.Body, fr.Channel)
+				continue
 			}
+			// we MUST remove the remote channel from our map as soon as we receive
+			// the ack (i.e. before passing it on to the session mux) on the session
+			// ending since the numbers are recycled.
+			delete(sessionsByRemoteChannel, fr.Channel)
+
+		default:
+			// pass on performative to the correct session
+			session, ok = sessionsByRemoteChannel[fr.Channel]
+			if !ok {
+				err = fmt.Errorf("%T: didn't find channel %d in sessionsByRemoteChannel", fr.Body, fr.Channel)
+				continue
+			}
+		}
 
-		// connection is complete
-		case <-c.closeMux:
+		select {
+		case session.rx <- fr:
```

Comment:
I'd be interesting to see how often we stall here. The 'session.rx' is an unbuffered channel so falling behind in one session could mean blocking the read loop altogther. Am I reading that incorrectly?

(also this is not a new problem, but it's just more clear how the code works since you stripped away more code!)

Reply:
Your reading is correct and is something I've been pondering on for some time now.  Unfortunately, removing the session mux is much more complicated than conn's mux as all of the flow control is managed there and depends on proper synchronization of incoming and outgoing frames (I've taken a few stabs at it, it's not pretty).